### PR TITLE
More Events: Add More Events extension

### PR DESCRIPTION
**Pull request by LilyMakesThings**
I've wanted to create this extension for a long while, this is about as many blocks as I can think of.

I needed this extension for specifically the "broadcast to target" block. (In hindsight, it may need some work).

image

### CODE REVIEW

**Comment by LilyMakesThings** [comment]
This block now exists thanks to @CST1229 
image

**Comment by LilyMakesThings** [comment]
Thumbnail:
image

**Comment by BlueDome77** [comment]
"when extension added" will pair nicely with ShovelUtils "load extension" block!

**Comment by BlueDome77** [comment]
Are you going to add this block?

scratchblocks

**Comment by LilyMakesThings** [comment]
> Are you going to add this block?
> 
> scratchblocks
> 
> It only makes sense cause the original "broadcast ()" block has a STRING input

That's impossible to add.

**Comment by BlueDome77** [comment]
> > Are you going to add this block?
> > scratchblocks
> > It only makes sense cause the original "broadcast ()" block has a STRING input
> 
> That's impossible to add.

why?

**Comment by LilyMakesThings** [comment]
I forgot to say but I made the changes to this block I eluded to:
- It now supports sending messages to the stage target
- The menu isn't broken 
- It sends to clones as well
image

**Comment by LilyMakesThings** [comment]
A few questions:
- Are merge conflicts something I'll need to resolve?
- Should these blocks be removed?
image

**Comment by OeildeLynx31** [comment]
> A few questions:
> - Are merge conflicts something I'll need to resolve?
> - Should these blocks be removed?
> image
> 

For the second question, I think you can add a menu instead of the string input, but in which we can drag reporters. The menu could have all the basics events, like 'green flag', or 'stop' ...

**Comment by SurvExe1Pc** [comment]
> > A few questions:
> > 
> > * Are merge conflicts something I'll need to resolve?
> > * Should these blocks be removed?
> >   image
> 
> For the second question, I think you can add a menu instead of the string input, but in which we can drag reporters. The menu could have all the basics events, like 'green flag', or 'stop' ...

i think its like that so you can start extension hats also

**Comment by NexusKitten** [comment]
looking forward to this extension!

**Comment by LilyMakesThings** [comment]
> I'm confused...
> 
> Screenshot 2023-07-11 at 8 59 18 AM Screenshot 2023-07-11 at 9 01 55 AM Source: [Penguin Mod](https://studio.penguinmod.site/editor.html)

I still don't quite know why I was credited, I think it was either because of the idea or because of the one block they share with this extension. Don't pay attention to penguinmod, this is the turbowarp repository

File: extensions/Lily/MoreEvents.js [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+    { text: '3', value: '3' },
+    { text: '4', value: '4' },
+    { text: '5', value: '5' },
+    { text: '6', value: '6' },
+    { text: '7', value: '7' },
+    { text: '8', value: '8' },
+    { text: '9', value: '9' },
+  ];
+
+  setInterval(() => {
```
Comment (GarboMuffin) [inline extensions/Lily/MoreEvents.js:73]: setInterval like this significantly impacts battery life. use vm.on('BEFORE_STEP', () => {}) instead (from memory, it's something like that, animated text uses it)

File: extensions/Lily/MoreEvents.js [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+        runtime.shouldExecuteStopClicked = true;
+      });
+      const originalGreenFlag = vm.greenFlag;
+      vm.greenFlag = function() {
+        runtime.shouldExecuteStopClicked = false;
+        originalGreenFlag.call(this);
+      };
+
+      runtime.on('STAGE_SIZE_CHANGED', () => {
+        runtime.startHats('lmsMoreEvents_whenRuntimeOptionChanged', {
```
Comment (GarboMuffin) [inline extensions/Lily/MoreEvents.js:106]: should these not be in runtime options

File: extensions/Lily/MoreEvents.js (hunk 1 above)
Comment (GarboMuffin) [reply extensions/Lily/MoreEvents.js:73]: its BEFORE_EXECUTE, the stop sign code below already uses it

File: extensions/Lily/MoreEvents.js (hunk 2 above)
Comment (LilyMakesThings) [reply extensions/Lily/MoreEvents.js:106]: I can remove them and put them into a separate PR later, if that's preferred. I'm inclined to agree

**Comment by LilyMakesThings** [comment]
I removed "when runtime options changed" and "when extension added", but kept the turbo mode block.

**Comment by LilyMakesThings** [comment]
Man I don't even know how to resolve conflicts myself 😭

**Comment by LilyMakesThings** [comment]
Added data sending through broadcast messages
image

**Comment by LilyMakesThings** [comment]
Using XML blocktype to get the broadcast menu into the blocks. Doing this until there's a way to do it without XML
image

**Comment by GarboMuffin** [comment]
There is a VM change that is merged but not deployed yet that will make whenValueChanged work properly

**Comment by LilyMakesThings** [comment]
should "forever" only run while the project is active?

**Comment by LilyMakesThings** [comment]
All of these have now been implemented. I'm not adding anymore than this.

image

**Comment by LilyMakesThings** [comment]
...except that.

**Comment by GarboMuffin** [comment]
is this done?

**Comment by LilyMakesThings** [comment]
yep

**Comment by LilyMakesThings** [comment]
OH DAMN

**Pull request merged**